// sdk-triggerx/dist/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerXClient = void 0;
const axios_1 = __importDefault(require("axios"));
class TriggerXClient {
    constructor(apiKey, config) {
        this.apiKey = apiKey; // Initialize the apiKey
        // const baseConfig = getConfig();
        this.client = axios_1.default.create({
            baseURL: 'https://data.triggerx.network', //'http://localhost:9002', //'https://data.triggerx.network',
            headers: { 'Authorization': `Bearer ${this.apiKey}` }, // Set the API key here
            timeout: 30000, // 30 second timeout
            ...config,
        });
    }
    // Method to get the API key
    getApiKey() {
        return this.apiKey;
    }
    async get(url, config) {
        const response = await this.client.get(url, config);
        return response.data;
    }
    async post(url, data, config) {
        const response = await this.client.post(url, data, config);
        return response.data;
    }
    // New PUT method
    async put(url, data, config) {
        const response = await this.client.put(url, data, config);
        return response.data;
    }
}
exports.TriggerXClient = TriggerXClient;
